libraries/chain/include/eosio/chain/abi_serializer.hpp

```diff
@@ -577,32 +577,33 @@ namespace impl {
          const variant_object& vo = v.get_object();
          EOS_ASSERT(vo.contains("signatures"), packed_transaction_type_exception, "Missing signatures");
          EOS_ASSERT(vo.contains("compression"), packed_transaction_type_exception, "Missing compression");
-         from_variant(vo["signatures"], ptrx.signatures);
-         from_variant(vo["compression"], ptrx.compression);
+         std::vector<signature_type> signatures;
+         packed_transaction::compression_type compression;
+         from_variant(vo["signatures"], signatures);
+         from_variant(vo["compression"], compression);
 
-         // TODO: Make this nicer eventually. But for now, if it works... good enough.
+         bytes packed_cfd;
```

Comment:
The `std::vector<bytes> cfd;` could also be moved up here, and a `bool use_packed_cfd = true;` also added here.

The code that checks for `packed_context_free_data` or `context_free_data` fields and extracts the appropriate one from the variant is repeated twice below. It could also be pulled out and just included once here. If there is no valid `packed_context_free_data` to extract, but there is a `context_free_data` field, it would set the `use_packed_cfd` to `false`.

Reply:
#6494